.grit/patterns/jest_to_vitest.md

```diff
@@ -0,0 +1,149 @@
+---
+title: Jest to Vitest
+---
+
+Convert Jest test to Vitest.
+
+tags: #migration, #js
+
+```grit
+engine marzano(0.1)
+language js
+
+pattern updated_imports() {
+  call_expression($function) where {
+    $function <: or {
+      js"test",
+      js"expect",
+      js"describe",
+      js"it",
+      js"beforeEach",
+      js"afterEach",
+      js"beforeAll",
+      js"afterAll"
+    } where {
+      $source = `'vitest'`,
+      $function <: ensure_import_from($source),
+      $vi = `vi`,
+      $vi <: ensure_import_from($source)
+    }
+  }
+}
+
+or {
+  // Imports
+  updated_imports(),
+  `import $_ from 'jest'` => .,
+
+  `jest.requireActual($module)` => `await vi.importActual($module)`,
+
+  `JEST_WORKER_ID` => `VITEST_POOL_ID`,
+
+  // module mocks
+  `jest.mock($module, $mockImplementation)` where { 
+    $mockImplementation <: arrow_function($body) where { 
+      $body <: literal_value() => `({ 
+        default: $body 
+      })`
+    }
+  } => `vi.mock($module, $mockImplementation)`,
+  `jest.mock` => `vi.mock`,
+
+  // done callback
+  `$sym($description, $callback)` where {
+    and {
+        $sym <: contains {
+            or { `it`, `test`}
+        },
+        $callback <: arrow_function($body, $parameters, $parameter) where {
+            or {
+                // For matching `done => { ... }`
+                ! $parameter <: .,
+                // For matching `(done) => { ... }`
+                $parameters <: contains r"\w+",
+            }
+        } => `() => new Promise($callback)`,
+    }
+  } => `$sym($description, $callback)`,
+
+  // beforeAll, beforeEach, afterAll, afterEach hooks
+  `$sym($callback)` where {
+    and {
+      $sym <: contains {
+            or { `beforeAll`, `beforeEach`, `afterAll`, `afterEach`}
+        },
+      $callback <: arrow_function($body, $parameters, $parameter) where {
+        $body <: call_expression() => `{ $body }`
+      }
```

Comment:
Again no guarantee these are always arrow functions.

Reply:
Thanks @morgante I think with `function() { .. }` usages, they are already in good shape. WDYT?

Other thing is that, in ticket desc, you write `Ensures beforeAll and afterAll hooks` but in [followed link](https://vitest.dev/guide/migration.html#hooks) they write `beforeAll/beforeEach hooks may return teardown function in Vitest`. 
Could you please confirm that we should handle `beforeAll and afterAll` or `beforeAll and beforeEach` or all of them?

Reply:
You should handle all hooks. They have the same behaviors.